File: dsl/kiw.go
package dsl

import (
	"os"
	"regexp"
	"strings"

	"github.com/krewire/libs/markdown"
	"gopkg.in/yaml.v3"
)

var (
	styleRe    = regexp.MustCompile(`(?is)<style([^>]*)>(.*?)</style>`)
	scriptRe   = regexp.MustCompile(`(?is)<script([^>]*)>(.*?)</script>`)
	markdownRe = regexp.MustCompile(`(?is)<markdown[^>]*>(.*?)</markdown>`)
	attrRe     = regexp.MustCompile("([a-zA-Z_:][-a-zA-Z0-9_:.]*)(?:\\s*=\\s*(?:\"([^\"]*)\"|'([^']*)'|([^\\s\"'=<>`]+)))?")
)

// StyleBlock holds a scoped style block with its attributes.
type StyleBlock struct {
	Lang    string `json:"lang"`
	Scoped  bool   `json:"scoped"`
	Content string `json:"content"`
}

// ScriptBlock holds a script block with tier attributes (FRK-DSL-030/031).
// Lang is js|ts|go|rust, Hydrate is load|idle|visible, Server/Compute flags.
type ScriptBlock struct {
	Lang    string `json:"lang"`
	Hydrate string `json:"hydrate"`
	Server  bool   `json:"server"`
	Compute bool   `json:"compute"`
	Content string `json:"content"`
}

// KiwModule is the parsed result of a .kiw file.
// It is JSON-serializable and intentionally mirrors the JS parser output
// so the same .kiw file can be consumed from Go (html/template) and from
// JS/TS (string templates) without a custom toolchain.
type KiwModule struct {
	Frontmatter  map[string]any `yaml:",inline" json:"frontmatter"`
	Body         string         `json:"body"`
	Styles       []string       `json:"styles"`
	Scripts      []string       `json:"scripts"`
	StyleBlocks  []StyleBlock   `json:"styleBlocks"`
	ScriptBlocks []ScriptBlock  `json:"scriptBlocks"`
	Markdown     []string       `json:"markdown"`
	Raw          string         `json:"-"`
}

func parseAttrs(tag string) map[string]string {
	m := map[string]string{}
	for _, sm := range attrRe.FindAllStringSubmatch(tag, -1) {
		if len(sm) < 2 {
			continue
		}
		key := strings.ToLower(sm[1])
		val := ""
		if len(sm) > 2 && sm[2] != "" {
			val = sm[2]
		} else if len(sm) > 3 && sm[3] != "" {
			val = sm[3]
		} else if len(sm) > 4 && sm[4] != "" {
			val = sm[4]
		}
		m[key] = val
	}
	return m
}

// ParseKiw parses a .kiw file content into a KiwModule.
//
// Format (Astro-like, but YAML frontmatter for Go/JS native):
//
//	---
//	title: Landing
//	layout: Base
//	---
//	<h1>{{.Title}}</h1>
//	<style>h1{color:red}</style>
//	<script>console.log(1)</script>
//
// Frontmatter is optional YAML between leading ---\n ... ---\n.
// Body is html/template source with zero or more top-level <style> and <script>
// blocks extracted as scoped CSS / client JS.
func ParseKiw(src string) (*KiwModule, error) {
	m := &KiwModule{
		Frontmatter: map[string]any{},
		Raw:         src,
	}
	body := src

	if strings.HasPrefix(strings.TrimSpace(src), "---") {
		trimmed := strings.TrimLeft(src, "\r\n\t ")
		if strings.HasPrefix(trimmed, "---") {
			rest := trimmed[3:]
			// find closing ---\n
			idx := strings.Index(rest, "\n---")
			if idx >= 0 {
				fmRaw := rest[:idx]
				body = rest[idx+4:]
				// strip leading newline after closing ---
				body = strings.TrimLeft(body, "\r\n")
				var fm map[string]any
				if err := yaml.Unmarshal([]byte(fmRaw), &fm); err == nil && fm != nil {
					m.Frontmatter = fm
				}
			}
		}
	}

	styles := styleRe.FindAllStringSubmatch(body, -1)
	for _, sm := range styles {
		if len(sm) < 3 {
			continue
		}
		attrs := parseAttrs(sm[1])
		content := strings.TrimSpace(sm[2])
		m.Styles = append(m.Styles, content)
		lang := attrs["lang"]
		if lang == "" {
			lang = "css"
		}
		m.StyleBlocks = append(m.StyleBlocks, StyleBlock{
			Lang:    strings.ToLower(lang),
			Scoped:  false,
			Content: content,
		})
		if _, ok := attrs["scoped"]; ok {
			m.StyleBlocks[len(m.StyleBlocks)-1].Scoped = true
		}
	}
	body = styleRe.ReplaceAllString(body, "")

	scripts := scriptRe.FindAllStringSubmatch(body, -1)
	for _, sm := range scripts {
		if len(sm) < 3 {
			continue
		}
		attrs := parseAttrs(sm[1])
		content := strings.TrimSpace(sm[2])
		m.Scripts = append(m.Scripts, content)
		lang := strings.ToLower(attrs["lang"])
		hydrate := strings.ToLower(attrs["hydrate"])
		_, hasServer := attrs["server"]
		_, hasCompute := attrs["compute"]
		if lang == "" {
			lang = "js"
		}
		if lang == "js" || lang == "ts" || lang == "go" {
			if !hasServer && !hasCompute && hydrate == "" {
				hydrate = "load"
			}
		}
		m.ScriptBlocks = append(m.ScriptBlocks, ScriptBlock{
			Lang:    lang,
			Hydrate: hydrate,
			Server:  hasServer,
			Compute: hasCompute,
			Content: content,
		})
	}
	body = scriptRe.ReplaceAllString(body, "")

	markdowns := markdownRe.FindAllStringSubmatch(body, -1)
	for _, mm := range markdowns {
		if len(mm) > 1 {
			m.Markdown = append(m.Markdown, strings.TrimSpace(mm[1]))
		}
	}
	body = markdownRe.ReplaceAllStringFunc(body, func(match string) string {
		sub := markdownRe.FindStringSubmatch(match)
		if len(sub) < 2 {
			return ""
		}
		inner := strings.TrimSpace(sub[1])
		if inner == "" {
			return ""
		}
		rendered, _ := markdown.Render([]byte(inner))
		return "\n" + strings.TrimSpace(rendered) + "\n"
	})

	m.Body = strings.TrimSpace(body)
	return m, nil
}

// ParseKiwFile reads and parses a .kiw file at path.
func ParseKiwFile(path string) (*KiwModule, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return ParseKiw(string(b))
}

File: dsl/kiw.ts
/**
 * Kiw DSL — JS/TS parser (mirrors Go ssg/kiw.go)
 *
 * DSL:
 * ---
 * title: Landing
 * layout: Base
 * ---
 * <h1>{{title}}</h1>
 * <style>h1{color:red}</style>
 * <script>console.log(1)</script>
 *
 * Frontmatter is YAML (parse with `yaml` npm package or `js-yaml`),
 * body is HTML with Go html/template `{{.Title}}` or JS `{title}`.
 * For JS, treat body as string template and replace {{.X}} / {{X}} / {X}.
 */

export interface StyleBlock { lang: string; scoped: boolean; content: string }
export interface ScriptBlock { lang: string; hydrate: string; server: boolean; compute: boolean; content: string }

export interface KiwModule {
  frontmatter: Record<string, any>
  body: string
  styles: string[]
  scripts: string[]
  styleBlocks: StyleBlock[]
  scriptBlocks: ScriptBlock[]
  markdown: string[]
  raw: string
}

const styleRe = /<style([^>]*)>([\s\S]*?)<\/style>/gi
const scriptRe = /<script([^>]*)>([\s\S]*?)<\/script>/gi
const markdownRe = /<markdown[^>]*>([\s\S]*?)<\/markdown>/gi
const attrRe = /([a-zA-Z_:][-a-zA-Z0-9_:.]*)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s"'=<>`]+)))?/g

function parseAttrs(tag: string): Record<string, string> {
  const out: Record<string, string> = {}
  let m: RegExpExecArray | null
  attrRe.lastIndex = 0
  while ((m = attrRe.exec(tag)) !== null) {
    const key = m[1].toLowerCase()
    const val = m[2] ?? m[3] ?? m[4] ?? ""
    out[key] = val
  }
  return out
}

export function parseKiw(src: string): KiwModule {
  let frontmatter: Record<string, any> = {}
  let body = src
  const trimmed = src.trimStart()
  if (trimmed.startsWith("---")) {
    const rest = trimmed.slice(3)
    const idx = rest.indexOf("\n---")
    if (idx >= 0) {
      const fmRaw = rest.slice(0, idx)
      body = rest.slice(idx + 4).replace(/^\r?\n/, "")
      try {
        // minimal YAML parse without deps for MVP: key: value lines
        frontmatter = parseYamlMinimal(fmRaw)
      } catch {}
    }
  }

  const styles: string[] = []
  const styleBlocks: StyleBlock[] = []
  body = body.replace(styleRe, (_m, attrs, inner) => {
    const a = parseAttrs(attrs)
    const content = inner.trim()
    styles.push(content)
    const lang = (a["lang"] || "css").toLowerCase()
    styleBlocks.push({ lang, scoped: "scoped" in a, content })
    return ""
  })

  const scripts: string[] = []
  const scriptBlocks: ScriptBlock[] = []
  body = body.replace(scriptRe, (_m, attrs, inner) => {
    const a = parseAttrs(attrs)
    const content = inner.trim()
    scripts.push(content)
    let lang = (a["lang"] || "js").toLowerCase()
    let hydrate = (a["hydrate"] || "").toLowerCase()
    const server = "server" in a
    const compute = "compute" in a
    if ((lang === "js" || lang === "ts" || lang === "go") && !server && !compute && !hydrate) hydrate = "load"
    scriptBlocks.push({ lang, hydrate, server, compute, content })
    return ""
  })

  const markdown: string[] = []
  body = body.replace(markdownRe, (_m, inner) => {
    markdown.push(inner.trim())
    // Keep markdown position by injecting raw markdown (Go side renders to HTML)
    // JS consumers can render with a markdown lib if needed.
    return "\n" + inner.trim() + "\n"
  })

  return {
    frontmatter,
    body: body.trim(),
    styles,
    scripts,
    styleBlocks,
    scriptBlocks,
    markdown,
    raw: src,
  }
}

function parseYamlMinimal(src: string): Record<string, any> {
  const out: Record<string, any> = {}
  for (const line of src.split("\n")) {
    const t = line.trim()
    if (!t || t.startsWith("#")) continue
    const colon = t.indexOf(":")
    if (colon === -1) continue
    const k = t.slice(0, colon).trim()
    const v = t.slice(colon + 1).trim().replace(/^["']|["']$/g, "")
    out[k] = v
  }
  return out
}

export function parseKiwFile(src: string): KiwModule {
  return parseKiw(src)
}

// Example (native JS):
// import { parseKiw } from "./kiw.ts"
// const mod = parseKiw(await Deno.readTextFile("pages/index.kiw"))
// console.log(mod.frontmatter.title) // "Landing"
// document.body.innerHTML = mod.body.replace("{{.Title}}", mod.frontmatter.title)
